indexer/osm_editor.cpp

```diff
@@ -678,21 +690,31 @@ NewFeatureCategories Editor::GetNewFeatureCategories() const
   return res;
 }
 
-namespace
+FeatureID Editor::GenerateNewFeatureId(MwmSet::MwmId const & id)
 {
-FeatureID GenerateNewFeatureId(FeatureID const & oldFeatureId)
-{
-  // TODO(AlexZ): Stable & unique features ID generation.
-  // TODO(vng): Looks like new feature indexes should uninterruptedly continue after existing indexes in mwm file.
-  static uint32_t newIndex = 0x0effffff;
-  return FeatureID(oldFeatureId.m_mwmId, newIndex++);
+  // TODO(vng): Double-check if new feature indexes should uninterruptedly continue after existing indexes in mwm file.
+  uint32_t featureIndex = kStartIndexForCreatedFeatures;
+  auto const found = m_features.find(id);
+  if (found != m_features.end())
+  {
+    // Scan all already created features and choose next available ID.
+    for (auto const & feature : found->second)
+    {
+      if (feature.second.m_status == FeatureStatus::Created && featureIndex <= feature.first)
+        featureIndex = feature.first + 1;
+    }
```

Comment:
Because only one ui thread at a time can create new features.

> On 9.3.2016, at 11.20, ygorshenin <notifications@github.com> wrote:
> 
> In indexer/osm_editor.cpp:
> 
> > -{
> > -  // TODO(AlexZ): Stable & unique features ID generation.
> > -  // TODO(vng): Looks like new feature indexes should uninterruptedly continue after existing indexes in mwm file.
> > -  static uint32_t newIndex = 0x0effffff;
> > -  return FeatureID(oldFeatureId.m_mwmId, newIndex++);
> > +  // TODO(vng): Double-check if new feature indexes should uninterruptedly continue after existing indexes in mwm file.
> > +  uint32_t featureIndex = kStartIndexForCreatedFeatures;
> > +  auto const found = m_features.find(id);
> > +  if (found != m_features.end())
> > +  {
> > +    // Scan all already created features and choose next available ID.
> > +    for (auto const & feature : found->second)
> > +    {
> > +      if (feature.second.m_status == FeatureStatus::Created && featureIndex <= feature.first)
> > +        featureIndex = feature.first + 1;
> > +    }
> This method is not thread-safe, right? Why?
> 
> —
> Reply to this email directly or view it on GitHub.
>